File: python/day01_01.py
def sum_calibration_values(filename):
    total_sum = 0

    with open(filename, 'r') as file:
        for line in file:
            first_digit = next((char for char in line if char.isdigit()), None)
            last_digit = next(
                (char for char in reversed(line) if char.isdigit()), None)

            if first_digit and last_digit:
                calibration_value = int(first_digit + last_digit)
                total_sum += calibration_value

    return total_sum


filename = "../input"
total_sum = sum_calibration_values(filename)
print(f"Total sum of calibration values: {total_sum}")

File: rust/day01_01.rs
use std::fs;

fn sum_calibration_values(document: &str) -> u32 {
    document
        .lines()
        .filter_map(|line| {
            let first_digit = line.chars().find(|c| c.is_digit(10));
            let last_digit = line.chars().rev().find(|c| c.is_digit(10));

            match (first_digit, last_digit) {
                (Some(f), Some(l)) => Some(f.to_digit(10).unwrap() * 10 + l.to_digit(10).unwrap()),
                _ => None,
            }
        })
        .sum()
}

fn main() {
    let document = fs::read_to_string("../input").expect("Failed to read 'input' file");

    let total_sum = sum_calibration_values(&document);
    println!("Total sum of calibration values: {}", total_sum);
}
